src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/asm/Disassembler.java

```diff
@@ -59,44 +64,57 @@ private Disassembler(long startPc, byte[] code) {
 
       // Lazily load hsdis
       if (decode_function == 0) {
-         StringBuilder path = new StringBuilder(System.getProperty("java.home"));
-         String sep = System.getProperty("file.separator");
-         String os = System.getProperty("os.name");
-         String libname = "hsdis";
-         String arch = System.getProperty("os.arch");
-         if (os.lastIndexOf("Windows", 0) != -1) {
-            if (arch.equals("x86")) {
-               libname +=  "-i386";
-            } else if (arch.equals("amd64")) {
-               libname +=  "-amd64";
-            } else {
-               libname +=  "-" + arch;
-            }
-            path.append(sep + "bin" + sep);
-            libname += ".dll";
-         } else if (os.lastIndexOf("Linux", 0) != -1) {
-            if (arch.equals("x86") || arch.equals("i386")) {
-               path.append(sep + "lib" + sep + "i386" + sep);
-               libname += "-i386.so";
-            } else if (arch.equals("amd64") || arch.equals("x86_64")) {
-               path.append(sep + "lib" + sep + "amd64" + sep);
-               libname +=  "-amd64.so";
-            } else {
-               path.append(sep + "lib" + sep + arch + sep);
-               libname +=  "-" + arch + ".so";
+         // 1. <home>/lib/<vm>/libhsdis-<arch>.so
+         // 2. <home>/lib/<vm>/hsdis-<arch>.so
+         // 3. <home>/lib/hsdis-<arch>.so
+         // 4. hsdis-<arch>.so  (using LD_LIBRARY_PATH)
+         Properties targetSysProps = VM.getVM().getSystemProperties();
+         String os = targetSysProps.getProperty("os.name");
+         String ext = ".so";
+         if (os.contains("Windows")) {
+            ext = ".dll";
+         } else if (os.contains("Mac OS")) {
+            ext = ".dylib";
+         }
+
+         String jvmPattern = "^(lib)?jvm\\" + ext + "$";
```

Comment:
A comment before this section would be useful. Something like "Find the full path to libjvm.so (jvm.dll and libjvm.dylib on Windows and OSX)."